(WIP) Subscription & events for Commonalities 0.5.0 (#313)
camaraproject/Commonalities: documentation/API-design-guidelines.md

          - [Examples](#examples)
        - [APIs which deal with explicit subscriptions](#apis-which-deal-with-explicit-subscriptions)
        - [API-level scopes (sometimes referred to as wildcard scopes in CAMARA)](#api-level-scopes-sometimes-referred-to-as-wildcard-scopes-in-camara)

>     - [11.7 API access restriction (Based on token \& client Id)](#117-api-access-restriction-based-on-token--client-id)

Reviewer: Align the section name

Author: Fixed

The scopes will always be those defined in the API Specs YAML files. Thus, a scope would only provide access to all endpoints and resources of an API if it is explicitly defined in the API Spec YAML file and agreed in the corresponding API subproject. 

> ### 11.7 API access restriction (Based on token & client Id)

Reviewer 2: I would prefer to have more general section title - if agree, then it should be modified also in ToC

```suggestion
### 11.7 Resource access restriction 
```

In some CAMARA APIs there are functions to create resource (via POST) and then later query them via id and/or list (with GET) or delete them (via DELETE). For example we have sessions, payments, subscriptions, etc..

For the GET and DELETE operations we must restrict the resource(s) targeted depending on the information provided in the request. Basically we consider 2 filters:
* API client (aka ClientId)
* access token

| Operation |  3-legged access token is used | 2-legged access token is used |
|-----------|--------------------------------|-------------------------------|

> | GET/{id} | - The resource queried must have been created for the end user associated with the access token. <br> - The resource queried must must have been created by the same API client given in the access token. | - The resource queried must have been created by the same API client given in the access token. |

Reviewer 3: ```suggestion
| GET/{id} | - The resource queried must have been created for the end user associated with the access token. <br> - The resource queried must have been created by the same API client given in the access token. | - The resource queried must have been created by the same API client given in the access token. |
```

> | GET/{id} | - The resource queried must have been created for the end user associated with the access token. <br> - The resource queried must must have been created by the same API client given in the access token. | - The resource queried must must have been created by the same API client given in the access token. |

Reviewer 3: ```suggestion
| GET/{id} | - The resource queried must have been created for the end user associated with the access token. <br> - The resource queried must must have been created by the same API client given in the access token. | - The resource queried must have been created by the same API client given in the access token. |
```

| GET/ | - Return all resource(s) created by the API consumer that are associated with both the end user identified by the access token and the same API client given in the access token. | - Return all resource(s) created by the same API client given in the access token. |

> | DELETE/{id} | - The resource to be deleted must have been created for the end user associated with the access token. <br> - The resource to be deleted must must have been created by the same API client given in the access token. | - The resource to be deleted must must have been created by the same API client given in the access token. |

Reviewer 3: ```suggestion
| DELETE/{id} | - The resource to be deleted must have been created for the end user associated with the access token. <br> - The resource to be deleted must have been created by the same API client given in the access token. | - The resource to be deleted must have been created by the same API client given in the access token. |
```

(e.g. the notification source for each of the supported events may be completely different,
in which case, separating the implementations is beneficial). 

> ##### Rules for subscriptions data minimization 

Reviewer 3: A placeholder added as comment to consider the generic advice hint mentioned here for event structure design :) https://github.com/camaraproject/Commonalities/issues/295#issuecomment-2418788007

Author: I've added the sentence: Application of data minimization design must be considered by the API working group for event structure definition.

These rules apply for subscription with device identifier
- If 3-legged access token is used, the POST and GET responses must not provide any device identifier.
- If 2-legged access token is used, the presence of a device identifier in the response is mandatory and should be the same identifier type than the one provided in the request.

> Application of data minimization design must be considered by the API working group for event structure definition.

Reviewer 3: ```suggestion
Application of data minimization design must be considered by the API Sub Project for event structure definition.
```

> Application of data minimization design must be considered by the API working group for event structure definition.

Author: We have discussed a bit this topic in Device Location sub projet. 

If we use the 2-legs the device identifier must be mandatory but then in the subscription answer and in the notification itself we can avoid to send this identifier. I proposed some suggestion here: https://github.com/camaraproject/DeviceLocation/issues/270 like adding a capability for the API consumer to indicate if the device & area information must be part of the event.

I think we probably cannot add more in commonalities for this release due to deadline but we can continue to explore the topic in API sub project. WDYT?

| initialEvent           | boolean            | Set to true by API consumer if consumer wants to get an event as soon as the subscription is created and current situation reflects event request. Example: Consumer request Roaming event. If consumer sets initialEvent to true and device is in roaming situation, an event is triggered. Up to API project decision to keep it. | optional    |

> Note on combined usage of initialEvent and subscriptionMaxEvents: 

Reviewer: ```suggestion
**Note** on combined usage of initialEvent and subscriptionMaxEvents: 
```

Unless explicitly decided otherwise by the API WG,  if an event is triggered following initialEvent set to `true`, this event will be counted towards subscriptionMaxEvents (if provided).
It is recommended to provide this clarification in all subscriptions API featuring subscriptionMaxEvents and initialEvent.

> Subscription status value table:

Reviewer: Suggestion: highlight **Subscription status value table:**

Author: Done

| initialEvent           | boolean            | Set to true by API consumer if consumer wants to get an event as soon as the subscription is created and current situation reflects event request. Example: Consumer request Roaming event. If consumer sets initialEvent to true and device is in roaming situation, an event is triggered. Up to API project decision to keep it. | optional    |

**Note** on combined usage of initialEvent and subscriptionMaxEvents: 

> Unless explicitly decided otherwise by the API WG,  if an event is triggered following initialEvent set to `true`, this event will be counted towards subscriptionMaxEvents (if provided).

Reviewer 2: Aligning with [Governance definitions](https://github.com/camaraproject/Governance/blob/main/ProjectStructureAndRoles.md#project-structure)

```suggestion
Unless explicitly decided otherwise by the API Sub Project,  if an event is triggered following initialEvent set to `true`, this event will be counted towards subscriptionMaxEvents (if provided).
```

> It is recommended to provide this clarification in all subscriptions API featuring subscriptionMaxEvents and initialEvent.

Reviewer 3: ```suggestion
It is recommended to provide this clarification in all subscription APIs featuring subscriptionMaxEvents and initialEvent.
```

**Note** on combined usage of initialEvent and subscriptionMaxEvents: 

> Unless explicitly decided otherwise by the API Sub Project,  if an event is triggered following initialEvent set to `true`, this event will be counted towards subscriptionMaxEvents (if provided).

Reviewer: ```suggestion
Unless explicitly decided otherwise by the API Sub Project, if an event is triggered following initialEvent set to `true`, this event will be counted towards subscriptionMaxEvents (if provided).
```

Reviewer: fix additional space

_Termination rules regarding subscriptionExpireTime & subscriptionMaxEvents usage_
* When client side providing a `subscriptionExpireTime` and/or `subscriptionMaxEvents` service side has to terminate the subscription without expecting a `DELETE` operation.

> * CAMARA did not impose limitations for `subscriptionExpireTime` or `subscriptionMaxEvents` but API providers may enforce limitations and must document them accordinly.

Reviewer 3: ```suggestion
* CAMARA does not impose limitations for `subscriptionExpireTime` or `subscriptionMaxEvents` but API providers may enforce limitations and must document them accordingly.
```

Note: For operational and troubleshooting purposes it is relevant to accommodate use of `x-correlator` header attribute. API listener implementations have to be ready to support and receive this data.

#### subscription-ends event
Specific event notification type "subscription-ends" is defined to inform listener about subscription termination.

> It is used when the `subscriptionExpireTime` or `subscriptionMaxEvents` has been reached, or, if the API server has to stop sending notification prematurely, or if the subscription request is managed asynchronously by the server and it is not able to provide the service. For this specific event, the `data` must feature `terminationReason` attribute. An enumeration of `terminationReason` is provided in event-subscription-template.yaml. 

Reviewer: Suggest to include relative link reference: to event-subscription-template.yaml

Author: Done

> It is used when the `subscriptionExpireTime` or `subscriptionMaxEvents` has been reached, or, if the API server has to stop sending notification prematurely, or if the subscription request is managed asynchronously by the server and it is not able to provide the service. For this specific event, the `data` must feature `terminationReason` attribute. An enumeration of `terminationReason` is provided in event-subscription-template.yaml (see in [Commonalities/artifacts/camara-cloudevents](/artifacts/camara-cloudevents) directory ``event-subscription-template.yaml``). 

Reviewer 3: ```suggestion
It is used when the `subscriptionExpireTime` or `subscriptionMaxEvents` has been reached, or, if the API server has to stop sending notifications prematurely, or if the subscription request is managed asynchronously by the server and it is not able to provide the service. For this specific event, the `data` must feature `terminationReason` attribute. An enumeration of `terminationReason` is provided in event-subscription-template.yaml (see in [Commonalities/artifacts/camara-cloudevents](/artifacts/camara-cloudevents) directory ``event-subscription-template.yaml``). 
```
